### PR TITLE
Enabling Global Illumination (GI VCT) for sensors in SDF

### PR DESCRIPTION
**Pull request by athenaz2**
# 🎉 New feature
## Summary

Global illumination (GI) has been enabled for the GUI with https://github.com/gazebosim/gz-sim/pull/1597, but not for the camera sensor view. This PR allows for GI to be enabled for sensors in the SDF as a `<global_illumination>` element within the `sensor_system` plugin. Currently supports the VCT technique, in the future can consider adding support for CI VCT.

Sensor view (left) and GUI (right) without GI VCT:

With GI VCT:

## Test it

[gi_for_sensors_demo.zip](https://github.com/user-attachments/files/16766562/gi_for_sensors_demo.zip)

Modify the `<global_illumination>` element in the SDF. This element is within the `sensor_system` plugin. The format is as follows:

```
<global_illumination type="vct">
    <enabled>true</enabled>
    <resolution>16 16 16</resolution>
    <octant_count>1 1 1</octant_count>
    <bounce_count>6</bounce_count>
    <high_quality>true</high_quality>
    <anisotropic>true</anisotropic>
    <thin_wall_counter>1.0</thin_wall_counter>
    <conserve_memory>true</conserve_memory>
    <debug_vis_mode>none</debug_vis_mode>
</global_illumination>
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/systems/sensors/Sensors.cc [hunk 1]
```diff
@@ -521,6 +628,52 @@ Sensors::~Sensors()
   this->dataPtr->Stop();
 }
 
+//TODO: why does math::Vector3d work but not math::Vector3i?
```
Comment (iche033) [inline src/systems/sensors/Sensors.cc:631]: looks like `math::Vector3i` is not one of the supported types: https://github.com/gazebosim/sdformat/blob/ba1bf8fedc0b6063d4312ad312e515e28f71cfb9/include/sdf/Param.hh#L529-L537

File: src/systems/sensors/Sensors.cc [hunk 2]
```diff
@@ -549,6 +702,54 @@ void Sensors::Configure(const Entity &/*_id*/,
...
+    {
+      gzerr << "Global illumination is only supported by the ogre2 render engine" << std::endl;
+    }
+    else
+    {
+      auto giElem = _sdf->FindElement("global_illumination");
+      std::string giType = giElem->GetAttribute("type")->GetAsString();
+      if (giType == "vct")
+      {
+        this->dataPtr->giVctParameters.enabled = giElem->Get<bool>("enabled", this->dataPtr->giVctParameters.enabled).first;
```
Comment (iche033) [inline src/systems/sensors/Sensors.cc:718]: wrap lines to 80char

File: src/systems/sensors/Sensors.cc [hunk 3]
```diff
@@ -521,6 +628,52 @@ Sensors::~Sensors()
...
+
+  _valueToSet[0] = static_cast<uint32_t>(parsedValues[0]);
+  _valueToSet[1] = static_cast<uint32_t>(parsedValues[1]);
+  _valueToSet[2] = static_cast<uint32_t>(parsedValues[2]);
+}
+
+static void ConvertDoubleToUInt32x3(uint32_t _valueToSet[3], math::Vector3d _defaultValue)
+{
+  ConvertDoubleToUInt32x3(nullptr, "", _valueToSet, _defaultValue);
+}
```
Comment (iche033) [inline src/systems/sensors/Sensors.cc:646]: Here's some slight tweaks to these helper functions:

* consider renaming the functions to better distinguish between the two
* rearranged some logic in the functions - I think they still provide the same results.
* used lower case for static functions
* made `math::Vector3d` argument const ref

```suggestion
static void convertVector3dToUInt32Array(uint32_t _valueToSet[3],
    const math::Vector3d &_vecValues)
{
  _valueToSet[0] = static_cast<uint32_t>(_vecValues[0]);
  _valueToSet[1] = static_cast<uint32_t>(_vecValues[1]);
  _valueToSet[2] = static_cast<uint32_t>(_vecValues[2]);
}

static void parseVector3dAsUInt32Array(sdf::ElementConstPtr _parentElem,
    const char *_childName, uint32_t _valueToSet[3],
    const math::Vector3d &_defaultValue)
{
  math::Vector3d parsedValues = (_parentElem == nullptr) ? _defaultValue :
      _parentElem->Get<math::Vector3d>(_childName, _defaultValue).first;

  convertVector3dToUInt32Array(_valueToSet, parsedValues);
}
```

Comment (iche033) [reply src/systems/sensors/Sensors.cc:646]: can you also document the arguments?

**Review by iche033 — COMMENTED** [review]
this generally looks good, added some suggestions

File: src/systems/sensors/Sensors.cc [hunk 4]
```diff
@@ -549,6 +702,54 @@ void Sensors::Configure(const Entity &/*_id*/,
...
+      auto giElem = _sdf->FindElement("global_illumination");
+      std::string giType = giElem->GetAttribute("type")->GetAsString();
+      if (giType == "vct")
+      {
+        this->dataPtr->giVctParameters.enabled = giElem->Get<bool>("enabled", this->dataPtr->giVctParameters.enabled).first;
+
+        if (giElem->HasElement("resolution"))
+          ConvertDoubleToUInt32x3(giElem, "resolution", this->dataPtr->giVctParameters.resolution, this->dataPtr->giDefaultData.resolution);
+        else 
+          ConvertDoubleToUInt32x3(this->dataPtr->giVctParameters.resolution, this->dataPtr->giDefaultData.resolution);
```
Comment (azeey) [inline src/systems/sensors/Sensors.cc:723]:
```suggestion

        this->dataPtr->giVctParameters.resolution = giElem->Get<math::Vector3<uint32_t>>("resolution", this->dataPtr->giDefaultData.resolution).first;
```

File: src/systems/sensors/Sensors.cc (hunk 2 above)
Comment (athenaz2) [reply src/systems/sensors/Sensors.cc]: Checked style with codecheck in f82c68e8b6a8e75ea5b01d2e168c7666f12b46f6

File: src/systems/sensors/Sensors.cc (hunk 3 above)
Comment (athenaz2) [reply src/systems/sensors/Sensors.cc]: Args documented in f82c68e8b6a8e75ea5b01d2e168c7666f12b46f6

File: src/systems/sensors/Sensors.cc [hunk 5]
```diff
@@ -549,6 +702,54 @@ void Sensors::Configure(const Entity &/*_id*/,
...
+    }
+    else
+    {
+      auto giElem = _sdf->FindElement("global_illumination");
+      std::string giType = giElem->GetAttribute("type")->GetAsString();
+      if (giType == "vct")
+      {
+        this->dataPtr->giVctParameters.enabled = giElem->Get<bool>("enabled", this->dataPtr->giVctParameters.enabled).first;
+
+        if (giElem->HasElement("resolution"))
```
[the remark this answers is not in the document]
Comment (athenaz2) [reply src/systems/sensors/Sensors.cc:737]: Added clarifying comment in f82c68e8b6a8e75ea5b01d2e168c7666f12b46f6

**Comment by iche033** [comment]
looks like there's an issue with GI on metal. Output from homebrew CI build:

```
281: [ RUN      ] CameraSensorGlobalIlluminationTest.GlobalIlluminationEnabled
281: libc++abi: terminating with uncaught exception of type Ogre::FileNotFoundException: OGRE EXCEPTION(6:FileNotFoundException): Cannot locate resource CrossPlatformSettings_piece_all.metal in resource group General or any other group. in ResourceGroupManager::openResource at /tmp/ogre2.3-20230918-86352-14va4s/ogre-next-2.3.1/OgreMain/src/OgreResourceGroupManager.cpp (line 793)
```

you can disable the test on mac using this [macro](https://github.com/gazebosim/gz-sim/blob/f600a441951c6ec0818008283b051a7770da1cd6/test/integration/camera_sensor_background.cc#L76). Can you ticket an issue on this in gz-rendering?

**Comment by athenaz2** [comment]
> looks like there's an issue with GI on metal. Output from homebrew CI build:
> 
> ```
> 281: [ RUN      ] CameraSensorGlobalIlluminationTest.GlobalIlluminationEnabled
> 281: libc++abi: terminating with uncaught exception of type Ogre::FileNotFoundException: OGRE EXCEPTION(6:FileNotFoundException): Cannot locate resource CrossPlatformSettings_piece_all.metal in resource group General or any other group. in ResourceGroupManager::openResource at /tmp/ogre2.3-20230918-86352-14va4s/ogre-next-2.3.1/OgreMain/src/OgreResourceGroupManager.cpp (line 793)
> ```
> 
> you can disable the test on mac using this [macro](https://github.com/gazebosim/gz-sim/blob/f600a441951c6ec0818008283b051a7770da1cd6/test/integration/camera_sensor_background.cc#L76). Can you ticket an issue on this in gz-rendering?

Okay, created an issue - https://github.com/gazebosim/gz-rendering/issues/1048. I also noted it in the issue description but the [GI unit test](https://github.com/gazebosim/gz-rendering/blob/gz-rendering8/test/common_test/GlobalIllumination_TEST.cc) also skips testing for Apple platform.

**Comment by azeey** [comment]
The test is still segfaulting. I'm wondering if it's an issue with having two rendering tests in the same file. Can you try commenting out the "GiNotEnabled" version to see if just running the test with GI enabled still segfaults?

File: tutorials/global_illumination.md [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+3) On the card, check the `Enabled` button to enable GI and change the parameter values as desired.
+
+### How to enable global illumination for the sensor
+
+GI can be enabled for sensors through the `<global_illumination>` element in the Sensors System plugin. VCT is supported for the sensor.
+
+#### Example usage with VCT
+
+We will demonstrate how to enable VCT for the sensor with the SDF file below. (The finished SDF file can be viewed [here](
+https://github.com/gazebosim/gz-sim/blob/main/examples/worlds/global_illumination.sdf).)
```
Comment (iche033) [inline tutorials/global_illumination.md:96]: change link to point to the `gz-sim8` branch

File: tutorials/global_illumination.md [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+3) On the card, check the `Enabled` button to enable GI and change the parameter values as desired.
+
+<div style="text-align:center;">
+    \image html files/global_illumination/gui_demo_with_vct.png width=60%
+</div>
+
+#### Example usage for CI VCT
+
+1) Open the [global_illumination.sdf](
+https://github.com/gazebosim/gz-sim/blob/main/examples/worlds/global_illumination.sdf) world using Vulkan with
```
Comment (iche033) [inline tutorials/global_illumination.md:79]: change link to point to the `gz-sim8` branch

File: tutorials/global_illumination.md [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+    * **Resolution**
+    * **Octant count**
+    * **Thin wall counter**
+    * **Area half size**
+
+## Enabling global illumination in Gazebo
+
+During simulation, the GUI and the camera sensor view are two different scenes. The GUI scene is rendered on the frontend client process, and the camera sensor scene is rendered on the backend server process. Thus, GI can be enabled on either or both of these processes.
+
+> **Note:** GI solutions require hardware that uses DX11 or GL4. <br>
```
Comment (iche033) [inline tutorials/global_illumination.md:48]: we haven't really tested it on DX11 before so let's not mention it for now until we verify it.

```suggestion
> **Note:** GI solutions require hardware that uses OpenGL4. <br>
```

File: tutorials/global_illumination.md [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+* **Bounce count**
+* **High quality**
+* **Anisotropic**
+* **Thin wall counter**
+* **Conserve memory**
+* **Debug visualization mode**
+
+### Cascaded Image Voxel Cone Tracing
+
+Compared to VCT, a scene rendered with CI VCT is slightly lower resolution, but the scene can be re-voxelized every frame, so it’s much faster when dealing with very large scenes. However, it is more memory-intensive than VCT.
```
Comment (iche033) [inline tutorials/global_illumination.md:31]:
```suggestion
Compared to VCT, a scene rendered with CI VCT is slightly lower quality, but the scene can be re-voxelized every frame, so it’s much faster when dealing with very large scenes. However, it is more memory-intensive than VCT.
```

File: tutorials/global_illumination.md [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+
+This tutorial showcases how to enable real-time global illumination in Gazebo.
+
+<div style="text-align:center;">
+    \image html files/global_illumination/global_illumination.gif width=60%
+</div>
+
+## About global illumination
+
+Global illumination (GI) techniques account for illumination from indirect light reflections. When GI is enabled, objects not only receive direct light from light sources, but also from indirect light bouncing off of other surfaces, ensuring nice reflections and soft shadows.
```
Comment (iche033) [inline tutorials/global_illumination.md:11]: technically shadows are a different technique from GI so suggestion to rephrase the last bit:

```suggestion
Global illumination (GI) techniques account for illumination from indirect light reflections. When GI is enabled, objects not only receive direct light from light sources, but also from indirect light bouncing off of other surfaces, ensuring nice reflections and a more natural appearance.
```

File: tutorials/global_illumination.md [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+> **Note:** CI VCT must be run with Vulkan as the render engine API backend.
+
+### How to enable global illumination for the GUI
+
+GI can be enabled for the GUI through a GUI plugin. Both VCT and CI VCT are supported for the GUI.
+
+#### Example usage for VCT
+
+1) Open the [global_illumination.sdf](
+https://github.com/gazebosim/gz-sim/blob/main/examples/worlds/global_illumination.sdf) world with
```
Comment (iche033) [inline tutorials/global_illumination.md:58]: update link to point to `gz-sim8` branch

File: test/worlds/camera_sensor_gi_enabled_true.sdf [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+      </global_illumination>
+    </plugin>
+    <plugin
+      filename="gz-sim-user-commands-system"
+      name="gz::sim::systems::UserCommands">
+    </plugin>
+    <!-- <plugin
+      filename="gz-sim-scene-broadcaster-system"
+      name="gz::sim::systems::SceneBroadcaster">
+    </plugin> -->
```
[the remark this answers is not in the document]
Comment (athenaz2) [reply test/worlds/camera_sensor_gi_enabled_true.sdf]: 95530a6aa1699b232512fdd76034cc279b2a730e

File: test/worlds/camera_sensor_gi_enabled_true.sdf [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+      <plugin filename="SelectEntities" name="Select Entities">
+          <gz-gui>
+              <property key="resizable" type="bool">false</property>
+              <property key="width" type="double">5</property>
+              <property key="height" type="double">5</property>
+              <property key="state" type="string">floating</property>
+              <property key="showTitleBar" type="bool">false</property>
+          </gz-gui>
+      </plugin>
+    </gui>
```
[the remark this answers is not in the document]
Comment (athenaz2) [reply test/worlds/camera_sensor_gi_enabled_true.sdf]: Whoops, was from creating SDF scene and ensuring cam lines up for test case. 95530a6aa1699b232512fdd76034cc279b2a730e

File: tutorials/global_illumination.md (hunk 6 above)
Comment (athenaz2) [reply tutorials/global_illumination.md]: 95530a6aa1699b232512fdd76034cc279b2a730e

File: tutorials/global_illumination.md (hunk 7 above)
Comment (athenaz2) [reply tutorials/global_illumination.md]: 95530a6aa1699b232512fdd76034cc279b2a730e

File: tutorials/global_illumination.md (hunk 11 above)
Comment (athenaz2) [reply tutorials/global_illumination.md]: 95530a6aa1699b232512fdd76034cc279b2a730e

**Comment by iche033** [comment]
I made a couple of minor changes in [f03476b](https://github.com/gazebosim/gz-sim/pull/2550/commits/f03476beeb39839b8c06521c4213c37f25afbd65) (removed white space), and [966f48c](https://github.com/gazebosim/gz-sim/pull/2550/commits/966f48cca0e8ea924cb3dba5b1f9f603320c4f8e) (hide gui plugins). This now looks good to me.

**Review by iche033 — APPROVED** [review]

**Review by ahcorde — APPROVED** [review]

**Comment by azeey** [comment]
This shouldn't have been merged since we are in code freeze. Let's make sure we don't forward port this to gz-sim9. Ideally, we won't need to forward port in the next few weeks, but if we do, we should just cherry-pick any necessary changes.

**Pull request merged**